fix(sbt-plugin): add new default registry URL

george-wilson-rea opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

## Changes

<!-- Describe what behavior is changed by this PR. -->

Change the default repository for SBT plugin lookup to `https://repo1.maven.org/maven2/`

## Context

<!-- Describe why you're making these changes if it's not already explained in a corresponding issue. -->
<!-- If you're closing an existing issue with this pull request, use the keyword Closes #issue_number. -->
<!-- If you're referencing an issue with this pull request, put it in a Markdown list like this: - #issue_number. -->

The bintray service has been sunset and now `https://repo.scala-sbt.org/scalasbt/` redirects to a mirror that no longer receives updates. Renovate should just default to maven.org for sbt plugins as it does for sbt packages. See this page for additional context: https://www.scala-sbt.org/1.x/docs/Bintray-For-Plugins.html


## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

viceice commented on lib/modules/datasource/sbt-plugin/index.ts:
```diff
@@ -9,10 +9,9 @@ import { SbtPackageDatasource } from '../sbt-package';
 
-export const SBT_PLUGINS_REPO =
-  'https://repo.scala-sbt.org/scalasbt/sbt-plugin-releases';
+export const SBT_MVN_REPO = 'https://repo1.maven.org/maven2';
```
how do you verified that all existing versions are migrated to maven.org?

george-wilson-rea replied:
Good point. I did not notice that older plugins aren't present in the newer repo. I've put the old URL back in, but I've made the maven URL the first in the list.

viceice reviewed (changes requested):
you need to verify that it still works with older plugins.

HonkingGoose commented on lib/modules/datasource/sbt-plugin/index.ts:
```diff
@@ -9,10 +9,9 @@ import { SbtPackageDatasource } from '../sbt-package';
 import { getLatestVersion, parseIndexDir } from '../sbt-package/util';
```
> Change the default repository for SBT plugin lookup to https://repo1.maven.org/maven2/

**Maintainers**: is this a `fix:` or `feat:` type change. It feels more like a `feat:` or maybe even a breaking feature change?

Unless the entire thing is broken right now because everybody migrated to the new stuff and Renovate is still using the old defaults. In that case it's probably a `fix:` type change.

george-wilson-rea replied:
I believe it's a `fix`. We run self-hosted Renovate and we find that sbt plugins have not been updated for some time. Recently we configured our `sbt-plugins` to use `https://repo1.maven.org/maven2/` and now we are getting some plugin updates.

According to [this page](https://www.scala-sbt.org/1.x/docs/Bintray-For-Plugins.html) (emphasis mine):

>As of 2021-04-18, we have migrated all sbt plugins and sbt 0.13 artifacts to the Artifactory instance, and redirected https://repo.scala-sbt.org/scalasbt/ to point to it as well, so existing builds should continue to work without making any changes today and after May 1st. **For plugin hosting, we will operate this as a read-only repository. Any new plugin releases should migrate to using [Sonatype OSS](https://www.scala-sbt.org/1.x/docs/Using-Sonatype.html)**.

So I think Renovate has not been updating sbt plugins since 2021, because it's been looking at a read-only mirror that was frozen in 2021 (`https://repo.scala-sbt.org/scalasbt/`).

When I go to Sonatype OSS, their releases repo `https://oss.sonatype.org/content/repositories/releases/` just redirects to `https://repo1.maven.org/maven2/`, so that seems like the right thing to configure.

HonkingGoose reviewed (commented):
Did you check if there are readme files related to this change that need to be updated?

rarkins reviewed (changes requested):
Needs at least one test on a "real" repository to make sure it works end to end. This is the type of change which would pass all unit tests while stopping plugin updating working altogether if something were wrong

george-wilson-rea commented:
@HonkingGoose When I search for "`sbt`" in the `docs` directory, I don't see anything that would need updating. Should I be looking anywhere else?

@rarkins How should I test this on a "real" repository? Could you please elaborate?

rarkins commented:
A public repo on GitHub.com which needs updates affected by this PR.

george-wilson-rea commented:
Here's a repo: https://github.com/george-wilson-rea/renovate-scala-plugins

In `plugins.sbt`, `sbt-scoverage` is set to `2.0.11`, which is not the latest version. I expect it to be updated to `2.0.12` but with recommended config, that doesn't happen.

george-wilson-rea commented:
Here's another repo, with an overridden sbt plugin config. Interestingly, sbt-scoverage is still not getting updated. I'm not sure what's going wrong here.
https://github.com/george-wilson-rea/renovate-scala-plugins-fixed

george-wilson-rea commented:
Interestingly, it looks like sbt plugins _are_ working by default, but for some reason `sbt-scoverage` specifically doesn't work with either setup. I might have misdiagnosed the problem here.
How can I investigate what's going wrong with `sbt-scoverage`?

HonkingGoose commented:
> @HonkingGoose When I search for "`sbt`" in the `docs` directory, I don't see anything that would need updating. Should I be looking anywhere else?

Hmm, the convention is that each manager/datasource/platform has a readme file. Maybe the sbt-plugin has no readme yet. It should have a readme though!

Please search in a similar directory for examples. 😉

HonkingGoose reviewed (changes requested):
Please add a readme to both the datasource and manager.

george-wilson-rea commented:
@HonkingGoose The sbt manager has a readme saying
```
Extracts Maven artifact dependencies for SBT.
```
which is accurate. I have added some minimal readmes to the `sbt-package` and `sbt-plugin` datasources. Is there anything else I should say in these documents?

george-wilson-rea commented:
I've started a discussion here with more detail about what seems to be going wrong: https://github.com/renovatebot/renovate/discussions/29755
I'm not sure this pull request will fix that problem but this pull request at least gives better defaults (new repo first, outdated repo second).

HonkingGoose commented on lib/modules/datasource/sbt-package/readme.md:
```diff
@@ -0,0 +1 @@
+This datasource finds SBT package updates from maven repositories.
```
```suggestion
This datasource finds SBT package updates from Maven repositories.
```

HonkingGoose commented on lib/modules/datasource/sbt-plugin/readme.md:
```diff
@@ -0,0 +1 @@
+This datasource finds SBT plugin updates from maven repositories.
```
```suggestion
This datasource finds SBT plugin updates from Maven repositories.
```

HonkingGoose reviewed (changes requested):
> ... I have added some minimal readmes to the `sbt-package` and `sbt-plugin` datasources. Is there anything else I should say in these documents?

It's usually good to mention the default settings in the readmes. So things like:

- By default, Renovate checks at human-readable-url for updates, but you can use `configOption=setting` to change this.
- By default, Renovate only finds updates when these conditions are met: list conditions.

HonkingGoose commented on lib/modules/datasource/sbt-plugin/index.spec.ts:
```diff
@@ -33,7 +33,7 @@ describe('modules/datasource/sbt-plugin/index', () => {
         .reply(404);
       httpMock
-        .scope('https://repo.maven.apache.org')
+        .scope('https://repo1.maven.org')
```
> I've started a discussion here with more detail about what seems to be going wrong: #29755
> I'm not sure this pull request will fix that problem but this pull request at least gives better defaults (new repo first, outdated repo second).

I'm marking this PR as `status:blocked`. Once we are ready to merge, a maintainer should:

- [ ] Remove the `status:blocked` label
- [ ] Resolve this review comment

george-wilson-rea commented:
I've now fixed my replication example and updated my bug report accordingly https://github.com/renovatebot/renovate/discussions/29755
This demonstrates that Renovate is not picking up plugin updates unless configured with extra registry URLs.

zharinov commented on lib/modules/datasource/sbt-plugin/index.ts:
```diff
@@ -9,10 +9,11 @@ import { SbtPackageDatasource } from '../sbt-package';
 import { getLatestVersion, parseIndexDir } from '../sbt-package/util';
 import type { GetReleasesConfig, ReleaseResult } from '../types';
 
+export const SBT_MVN_REPO = 'https://repo1.maven.org/maven2';
```
The `MAVEN_REPO` constant from the `maven` datasource should be used instead

viceice replied:
good catch

george-wilson-rea replied:
I've merged @zharinov 's pull request to address this.

zharinov commented:
We have to take into account different situations:
- Maven Central contains ancient plugin versions and `jfrog.io` contains the newer ones
- `jfrog.io` contains older versions, but newest ones are available in Maven Central

To address that, the `merge` registry strategy should be enabled.

viceice reviewed (commented):
see comment

zharinov commented:
Just opened PR with my suggestions: https://github.com/george-wilson-rea/renovate/pull/1

Because of `merge` registry strategy being used, now the exact order of registries doesn't really matter

george-wilson-rea commented on lib/modules/datasource/sbt-package/index.ts:
```diff
@@ -22,7 +26,7 @@ export class SbtPackageDatasource extends MavenDatasource {
   override readonly defaultVersioning = ivyVersioning.id;
 
-  override readonly registryStrategy = 'hunt';
+  override readonly registryStrategy: RegistryStrategy = 'hunt';
```
Now we've changed `sbt-plugin` to use the `merge` strategy - should `sbt-package` still be using `hunt` or should we change it to use `merge` too?

zharinov replied:
No, let's keep it as is, it should be fine

george-wilson-rea commented:
I've merged the PR from @zharinov and fixed up the tests. This should be unblocked now.

zharinov commented:
Now we don't need test changes, check it out:
- https://github.com/george-wilson-rea/renovate/pull/2

george-wilson-rea commented:
I've merged the latest suggestions from @zharinov, which has unfortunately dismissed the approving review. I've re-requested review per the contribution guidelines. Sorry for the churn.

HonkingGoose commented on lib/modules/datasource/sbt-package/readme.md:
```diff
@@ -0,0 +1,13 @@
+This datasource finds SBT package updates from Maven repositories.
+
+By default, Renovate checks <https://repo.maven.apache.org/maven2> for SBT packages. You can override this behavior by overriding the `registryUrls` setting. For example:
```
```suggestion
By default, Renovate checks `https://repo.maven.apache.org/maven2` for SBT packages.
You can override the default behavior with the `registryUrls` config option.
For example:
```

HonkingGoose commented on lib/modules/datasource/sbt-plugin/readme.md:
```diff
@@ -0,0 +1,14 @@
+This datasource finds SBT plugin updates from Maven repositories.
+
+By default, Renovate checks <https://repo1.maven.org/maven2/> for SBT plugins, and then falls back to the legacy URL <https://repo.scala-sbt.org/scalasbt/sbt-plugin-releases>. You can override this behavior by overriding the `registryUrls` setting. For example:
```
```suggestion
By default, Renovate:

1. Checks `https://repo1.maven.org/maven2/` for SBT plugins
1. If the above URL returns no results, then Renovate tries the _legacy_ URL: `https://repo.scala-sbt.org/scalasbt/sbt-plugin-releases`
 
You can override the default behavior with the `registryUrls` config option.
For example:
```

HonkingGoose reviewed (changes requested):
Some docs changes:

- Follow one sentence per line pattern
- Make a ordered list of the steps Renovate takes by default
- Small style fixes

george-wilson-rea commented:
Thanks for enabling `auto-merge` @rarkins . Could you, or someone else, please also approve so that this gets merged? Your last review was dismissed when I merged @zharinov 's suggestions.

viceice commented on lib/modules/datasource/sbt-plugin/index.spec.ts:
```diff
@@ -37,10 +38,12 @@ describe('modules/datasource/sbt-plugin/index', () => {
-            '<a href="scalatest_2.12/" title=\'scalatest_2.12/\'>scalatest_2.12/</a>\n' +
-            "<a href='scalatest_sjs2.12/'>scalatest_2.12/</a>" +
-            "<a href='scalatest_native2.12/'>scalatest_2.12/</a>",
+          codeBlock`
```
```suggestion
          html`
```

use `html` template 😉 

https://www.npmjs.com/package/common-tags#html

viceice commented on lib/modules/datasource/sbt-plugin/index.spec.ts:
```diff
@@ -52,74 +55,74 @@ describe('modules/datasource/sbt-plugin/index', () => {
-            '<pre><a href="scala_2.12/">scala_2.12/</a></pre>\n' +
-            '</body>\n' +
-            '</html>',
+          codeBlock`
```
```suggestion
          html`
```

viceice commented on lib/modules/datasource/sbt-plugin/index.spec.ts:
```diff
@@ -52,74 +55,74 @@ describe('modules/datasource/sbt-plugin/index', () => {
-            '<pre><a href="sbt_1.0/">sbt_1.0/</a></pre>\n' +
-            '</body>\n' +
-            '</html>\n',
+          codeBlock`
```
```suggestion
          html`
```

viceice commented on lib/modules/datasource/sbt-plugin/index.spec.ts:
```diff
@@ -52,74 +55,74 @@ describe('modules/datasource/sbt-plugin/index', () => {
-            '<pre><a href="0.5.5/">0.5.5/</a></pre>\n' +
-            '</body>\n' +
-            '</html>\n',
+          codeBlock`
```
```suggestion
          html`
```

viceice commented on lib/modules/datasource/sbt-plugin/index.spec.ts:
```diff
@@ -52,74 +55,74 @@ describe('modules/datasource/sbt-plugin/index', () => {
-            '<a href="sbt-coursier_2.12_1.0/">sbt-coursier_2.12_1.0/</a>\n' +
-            '<a href="sbt-coursier_2.12_1.0.0-M5/">sbt-coursier_2.12_1.0.0-M5/</a>\n' +
-            '<a href="sbt-coursier_2.12_1.0.0-M6/">sbt-coursier_2.12_1.0.0-M6/</a>\n',
+          codeBlock`
```
```suggestion
          html`
```

viceice commented on lib/modules/datasource/sbt-plugin/index.spec.ts:
```diff
@@ -52,74 +55,74 @@ describe('modules/datasource/sbt-plugin/index', () => {
-            '<a href="2.0.0-RC6-1/">2.0.0-RC6-1/</a>\n' +
-            '<a href="2.0.0-RC6-2/">2.0.0-RC6-2/</a>\n' +
-            '<a href="2.0.0-RC6-6/">2.0.0-RC6-6/</a>\n',
+          codeBlock`
```
```suggestion
          html`
```

viceice commented on lib/modules/datasource/sbt-plugin/index.spec.ts:
```diff
@@ -1,3 +1,4 @@
+import { codeBlock } from 'common-tags';
```
```suggestion
import { codeBlock, html } from 'common-tags';
```

viceice reviewed (commented):
otherwise LGTM
